crates/astria-sequencer/src/genesis.rs

```diff
@@ -4,29 +4,157 @@ use astria_core::primitive::v1::{
 };
 use penumbra_ibc::params::IBCParameters;
 use serde::{
-    de::Error as _,
     Deserialize,
-    Deserializer,
+    Serialize,
 };
 
 /// The genesis state for the application.
-#[derive(Debug, Deserialize)]
+///
+/// Verified to only contain valid fields (right now, addresses that have the same base prefix
+/// as set in `GenesisState::address_prefixes::base`).
+///
+/// **NOTE:** The fields should not be publicly accessible to guarantee invariants. However,
+/// it's easy to just go along with this for now.
+#[derive(Clone, Debug, Deserialize, Serialize)]
+#[serde(try_from = "UncheckedGenesisState", into = "UncheckedGenesisState")]
 pub(crate) struct GenesisState {
+    pub(crate) address_prefixes: AddressPrefixes,
     pub(crate) accounts: Vec<Account>,
-    #[serde(deserialize_with = "deserialize_address")]
     pub(crate) authority_sudo_address: Address,
-    #[serde(deserialize_with = "deserialize_address")]
     pub(crate) ibc_sudo_address: Address,
-    #[serde(deserialize_with = "deserialize_addresses")]
     pub(crate) ibc_relayer_addresses: Vec<Address>,
     pub(crate) native_asset_base_denomination: String,
     pub(crate) ibc_params: IBCParameters,
-    #[serde(deserialize_with = "deserialize_assets")]
     pub(crate) allowed_fee_assets: Vec<asset::Denom>,
     pub(crate) fees: Fees,
 }
 
-#[derive(Debug, Deserialize)]
+#[derive(Debug, thiserror::Error)]
+// allow: this error is only seen at chain init and never after so perf impact of too large enum
+// variants is negligible
+#[allow(clippy::result_large_err)]
+pub(crate) enum VerifiyGenesisError {
```

Comment:
```suggestion
pub(crate) enum VerifyGenesisError {
```

Reply:
Hah, good catch. Auto-complete spread it everywhere. bbbe762